dev/PushNotifications/PushManager.cpp

```diff
@@ -4,11 +4,111 @@
 #include "pch.h"
 #include "PushManager.h"
 #include "PushManager.g.cpp"
+#include <winrt\windows.networking.pushnotifications.h>
+#include <winerror.h>
+#include <algorithm>
+#include <list>
+#include <iostream>
+#include <ChannelResult.h>
+#include <winerror.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+// MAX BACKOFF TIME IS 16MINUTES 
+#define MAX_BACKOFF_SECONDS 960
+
+// Error code - platform is not connected
+#define WNP_E_NOT_CONNECTED 0x880403E8L
+
+// Error code - platform reconnecting with wns
+#define WNP_E_RECONNECTING  0x880403E9L
+
+// Error code - Current bind operation failed because another bind operation in progress
+#define WNP_E_BIND_USER_BUSY 0x880403FEL
 
 namespace winrt::Microsoft::ProjectReunion::implementation
 {
+    std::list<winrt::guid> PushManager::remoteIdList;
+    std::mutex PushManager::mtx;
+    std::unique_lock<std::mutex> PushManager::lock (mtx, std::defer_lock); // Avoid locking during the constructor call
+
+    bool PushManager::isChannelRequestRetryable(const winrt::hresult& hrException)
+    {
+        switch (hrException)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+            // E_TIMEOUT
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING
+        case WNP_E_BIND_USER_BUSY:
+            return true;
+        default:
+            return false;
+        }
+    }
+
     Windows::Foundation::IAsyncOperationWithProgress<Microsoft::ProjectReunion::ChannelResult, Microsoft::ProjectReunion::ChannelResult> PushManager::CreateChannelAsync(winrt::guid remoteId)
     {
-        throw hresult_not_implemented();
+        winrt::Microsoft::ProjectReunion::ChannelResult channelResult{ nullptr };
+        auto progress{ co_await winrt::get_progress_token() };
+
+       GUID remoteIdAbiGuid = reinterpret_cast<GUID&>(remoteId);
+       if (remoteIdAbiGuid == GUID_NULL)
```

Comment:
```suggestion
       if (remoteId == winrt::guid{})
```

Also nit: I don't remember [the specification](https://github.com/microsoft/ProjectReunion/issues/334) stating a null guid being invalid. I'm not against the check, just noting a tiny discrepancy.

Reply:
@jonwis https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/error-handling - "don't use exceptions for expected failures. Throw an exception only when an unexpected runtime error occurs" - based on the document I realized that when guid is null then we should not be throwing as this is an expected failure.  Can you please let me know if this needs to be considered a critical error instead and throw? Thanks.

Reply:
This is a critical developer usage error and you should throw.